// @balticcode/ngx-hotkeys/esm2015/lib/ngx-hotkeys.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Directive, ElementRef, Input } from '@angular/core';
import 'mousetrap';
import { NgxHotkeysService } from './ngx-hotkeys.service';
export class NgxHotkeysDirective {
    /**
     * @param {?} _hotkeysService
     * @param {?} _elementRef
     */
    constructor(_hotkeysService, _elementRef) {
        this._hotkeysService = _hotkeysService;
        this._elementRef = _elementRef;
        this._hotkeysList = [];
        this._oldHotkeys = [];
        this._mousetrapInstance = new Mousetrap(this._elementRef.nativeElement); // Bind hotkeys to the current element (and any children)
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        for (const hotkey of this.hotkeys) {
            /** @type {?} */
            const combo = Object.keys(hotkey)[0];
            /** @type {?} */
            const hotkeyObj = {
                combo: combo,
                handler: hotkey[combo]
            };
            /** @type {?} */
            const filtered = this._hotkeysService.get(combo);
            /** @type {?} */
            const oldHotkey = !!filtered && !!filtered.length ? filtered[0] : null;
            if (oldHotkey !== null) { // We let the user overwrite callbacks temporarily if you specify it in HTML
                // We let the user overwrite callbacks temporarily if you specify it in HTML
                this._oldHotkeys.push(oldHotkey);
                this._hotkeysService.unregister(oldHotkey);
            }
            this._hotkeysList.push(hotkeyObj);
            this._mousetrapInstance.bind(hotkeyObj.combo, hotkeyObj.handler);
        }
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        for (const hotkey of this._hotkeysList) {
            this._mousetrapInstance.unbind(hotkey.combo);
        }
        this._hotkeysService.register(this._oldHotkeys);
    }
}
NgxHotkeysDirective.decorators = [
    { type: Directive, args: [{
                selector: '[ngxHotkeys]',
                providers: [NgxHotkeysService]
            },] }
];
/** @nocollapse */
NgxHotkeysDirective.ctorParameters = () => [
    { type: NgxHotkeysService },
    { type: ElementRef }
];
NgxHotkeysDirective.propDecorators = {
    hotkeys: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    NgxHotkeysDirective.prototype.hotkeys;
    /** @type {?} */
    NgxHotkeysDirective.prototype._mousetrapInstance;
    /** @type {?} */
    NgxHotkeysDirective.prototype._hotkeysList;
    /** @type {?} */
    NgxHotkeysDirective.prototype._oldHotkeys;
    /** @type {?} */
    NgxHotkeysDirective.prototype._hotkeysService;
    /** @type {?} */
    NgxHotkeysDirective.prototype._elementRef;
}
