// @balticcode/ngx-hotkeys/esm5/lib/ngx-hotkeys.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { NgxCheatsheetComponent } from './ngx-cheatsheet/ngx-cheatsheet.component';
import { NgxHotkeysDirective } from './ngx-hotkeys.directive';
import { NgxHotkeysService } from './ngx-hotkeys.service';
import { HotkeyOptions } from './token';
var NgxHotkeysModule = /** @class */ (function () {
    function NgxHotkeysModule() {
    }
    /**
     * @param {?=} options
     * @return {?}
     */
    NgxHotkeysModule.forRoot = /**
     * @param {?=} options
     * @return {?}
     */
    function (options) {
        return {
            ngModule: NgxHotkeysModule,
            providers: [
                NgxHotkeysService,
                { provide: HotkeyOptions, useValue: options }
            ]
        };
    };
    NgxHotkeysModule.decorators = [
        { type: NgModule, args: [{
                    imports: [CommonModule],
                    declarations: [NgxCheatsheetComponent, NgxHotkeysDirective],
                    exports: [NgxHotkeysDirective, NgxCheatsheetComponent],
                    providers: [NgxHotkeysService]
                },] }
    ];
    return NgxHotkeysModule;
}());
export { NgxHotkeysModule };
