// @balticcode/ngx-hotkeys/esm5/lib/ngx-hotkeys.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Directive, ElementRef, Input } from '@angular/core';
import 'mousetrap';
import { NgxHotkeysService } from './ngx-hotkeys.service';
var NgxHotkeysDirective = /** @class */ (function () {
    function NgxHotkeysDirective(_hotkeysService, _elementRef) {
        this._hotkeysService = _hotkeysService;
        this._elementRef = _elementRef;
        this._hotkeysList = [];
        this._oldHotkeys = [];
        this._mousetrapInstance = new Mousetrap(this._elementRef.nativeElement); // Bind hotkeys to the current element (and any children)
    }
    /**
     * @return {?}
     */
    NgxHotkeysDirective.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        var e_1, _a;
        try {
            for (var _b = tslib_1.__values(this.hotkeys), _c = _b.next(); !_c.done; _c = _b.next()) {
                var hotkey = _c.value;
                /** @type {?} */
                var combo = Object.keys(hotkey)[0];
                /** @type {?} */
                var hotkeyObj = {
                    combo: combo,
                    handler: hotkey[combo]
                };
                /** @type {?} */
                var filtered = this._hotkeysService.get(combo);
                /** @type {?} */
                var oldHotkey = !!filtered && !!filtered.length ? filtered[0] : null;
                if (oldHotkey !== null) { // We let the user overwrite callbacks temporarily if you specify it in HTML
                    // We let the user overwrite callbacks temporarily if you specify it in HTML
                    this._oldHotkeys.push(oldHotkey);
                    this._hotkeysService.unregister(oldHotkey);
                }
                this._hotkeysList.push(hotkeyObj);
                this._mousetrapInstance.bind(hotkeyObj.combo, hotkeyObj.handler);
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
            }
            finally { if (e_1) throw e_1.error; }
        }
    };
    /**
     * @return {?}
     */
    NgxHotkeysDirective.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        var e_2, _a;
        try {
            for (var _b = tslib_1.__values(this._hotkeysList), _c = _b.next(); !_c.done; _c = _b.next()) {
                var hotkey = _c.value;
                this._mousetrapInstance.unbind(hotkey.combo);
            }
        }
        catch (e_2_1) { e_2 = { error: e_2_1 }; }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
            }
            finally { if (e_2) throw e_2.error; }
        }
        this._hotkeysService.register(this._oldHotkeys);
    };
    NgxHotkeysDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[ngxHotkeys]',
                    providers: [NgxHotkeysService]
                },] }
    ];
    /** @nocollapse */
    NgxHotkeysDirective.ctorParameters = function () { return [
        { type: NgxHotkeysService },
        { type: ElementRef }
    ]; };
    NgxHotkeysDirective.propDecorators = {
        hotkeys: [{ type: Input }]
    };
    return NgxHotkeysDirective;
}());
export { NgxHotkeysDirective };
if (false) {
    /** @type {?} */
    NgxHotkeysDirective.prototype.hotkeys;
    /** @type {?} */
    NgxHotkeysDirective.prototype._mousetrapInstance;
    /** @type {?} */
    NgxHotkeysDirective.prototype._hotkeysList;
    /** @type {?} */
    NgxHotkeysDirective.prototype._oldHotkeys;
    /** @type {?} */
    NgxHotkeysDirective.prototype._hotkeysService;
    /** @type {?} */
    NgxHotkeysDirective.prototype._elementRef;
}
